Update CPUManager stored state semantics

klueska opened the pull request:
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR updates the information that is stored as part of the CPUManager state. Instead of storing the CPUManager state keyed off of the `ContainerID`, we now store it keyed off of the combination of `PodUID`and `ContainerName`.  This change is necessary moving forward as we move to a model where we assign CPUs to containers at pod admit time, rather than container creation time.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The underlying format of the `CPUManager` state file has changed. Upgrades should be seamless, but any third-party tools that rely on reading the previous format need to be updated.
```

klueska committed:
Move containerMap out of static policy and into top-level CPUManager

klueska committed:
Change CPUManager state to key off of podUID and containerName

Previously, the state was keyed off of containerID intead of podUID and
containerName. Unfortunately, this is no longer possible as we move to a
to model where we we allocate CPUs to containers at pod adit time rather
than container start time.

This patch is the first step towards full migration to the new
semantics. Only the unit tests in cpumanager/state are passing. In
subsequent commits we will update the CPUManager itself to use these new
semantics.

This patch also includes code to do migration from the old checkpoint format
to the new one, assuming the existence of a ContainerMap with the proper
mapping of (containerID)->(podUID, containerName). A subsequent commit
will update code in higher layers to make sure that this ContainerMap is
made available to this state logic.

klueska committed:
Extend makePod() helper in CPUManager to take PodUID and ContainerName

klueska committed:
Fix bug in parsing int to string in CPUManager tests

klueska commented:
/assign @ConnorDoyle

klueska commented:
/test pull-kubernetes-integration

klueska commented:
/test pull-kubernetes-e2e-gce

mattjmcnaughton reviewed (commented):
I've been a little detached from the reviews on the CPUManager internals, so I'll leave technical specifics around this change to someone else :)

But, I do have one question. From reading the PR description, it looked like we plan to merge this diff to master before the subsequent necessary diff you mentioned is approved and ready to merge? If yes, that makes me a little nervous. If for some reason, the subsequent diff takes longer, it'd be unfortunate (I think? perhaps you can share more about the ramifications) to have the first diff in 1.17 but not the second.

klueska commented:
@mattjmcnaughton This patch should only be merged together with the patch to support upgrades. I'm working on this right now and will add it as a follow-on commit in this same PR once it's done, and update the description accordingly.

mattjmcnaughton commented:
> @mattjmcnaughton This patch should only be merged together with the patch to support upgrades. I'm working on this right now and will add it as a follow-on commit in this same PR once it's done, and update the description accordingly.

Gotcha, thanks @klueska ! Mind adding a `[WIP]` prefix to your PR title, so its clear to reviewers and doesn't get accidentally merged? Thanks!

derekwaynecarr commented:
@klueska this appears to take commits from the #84196 that is also in progress.  updating the title to WIP pending the core PRs merging first.

klueska committed:
Update CPUmanager containerMap to also return a containerRef

klueska committed:
Change CPUManager containerMap to key off of (podUID, containerName)

Previously it keyed off of a pointer to the actual pod / container,
which was unnecessary, and hard to work with (especially on the
retrieval side).

klueska committed:
Update CPUmanager containerMap to allow removal by containerRef

klueska committed:
Update CPUManager policies to adhere to new state semantics

klueska committed:
Update top-level CPUManager to adhere to new state semantics

For now, we just pass 'nil' as the set of 'initialContainers' for
migrating from old state semantics to new ones. In a subsequent commit
will we pull this information from higher layers so that we can pass it
down at this stage properly.

klueska commented:
@ConnorDoyle @derekwaynecarr 

This PR has now been rebased on top of master (well technically on top of #84525 which I'm hoping will merge soon).

However, we can't call this PR complete until we have a migration path of the old state format (keyed off of containerID) to the new state format (keyed off of podUID, containerName).

I'm curious what your thoughts are for how to ensure state stored by 1.16 kubelets will be readable by 1.17 kubelets, transformed to the new semantics, and then written back in the new format.

I imagine there must be some a sort of precedent on best practices for this, but my (admittedly limited) search for it hasn't turned up much.

klueska commented:
/test pull-kubernetes-e2e-gce-100-performance

klueska commented:
/test pull-kubernetes-kubemark-e2e-gce-big

cbf123 commented:
In OpenStack nova they had a mechanism that was basically opportunistic upgrades...that is, the code understood how to read both the old format and the new format, and when it read in the data in the old format it would convert the data to the new format and write it back.

I'm not sure it's needed here, but there was a separate "online upgrade" stage to ensure that all of the data had been updated to the new format.  This stage needed to be manually run before upgrading to a version that no longer supported the old data format.

Do we need to support fallback in case of a failed upgrade?  If so, it might make sense to use a different filename for the new saved state file, or maybe to write the new data to the state file in a way that the old version could still read.  In a future release you could delete the backwards-compatibility state information.

klueska commented:
The bigger issue is that we can’t immediately convert from one format to another as a „pre-step“ because we don’t have the pod uid or container name readily available as we are doing the recovery. We can do it lazily (which is probably what I’ll end up doing), but then it gets more complicated with keeping things in sync (not too bad though).

klueska commented:
Having an upgrade stage supported during kubelet startup would help a lot in this regard. Basically a state restore phase, followed by and upgrade phase before any components begin their normal startup.

As far as I know, no coordinated effort exists at present. All components just do their own checkpoint management as they come online, assuming no dependencies on any other components. Normally this is fine (you should checkpoint everything you need to recover in a component by component basis), but it’s problematic when upgrading to a format, where going forward you need additional info not in the old state file (as is the case here).

klueska committed:
Move CPUManager ContainerMap to its own package

klueska committed:
Move CPUManager Checkpoint restoration to Start() instead of New()

klueska committed:
Pass initial set of runtime containers to the CPUManager at startup

These information associatedd with these containers is used to migrate
the CPUManager state from it's old format to its new (i.e. keyed off of
podUID and containerName instead of containerID).

derekwaynecarr commented:
/assign

i will review this today.

derekwaynecarr commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -96,6 +97,10 @@ type manager struct {
 	// and the containerID of their containers
 	podStatusProvider status.PodStatusProvider
 
+	// containerMap provides a mapping from (pod, container) -> containerID
```
nit: its containerID -> (pod, container) per earlier commits

derekwaynecarr commented on pkg/kubelet/cm/cpumanager/state/checkpoint.go:
```diff
@@ -23,38 +23,87 @@ import (
+type CPUManagerCheckpointV2 struct {
+	PolicyName    string                       `json:"policyName"`
+	DefaultCPUSet string                       `json:"defaultCpuSet"`
+	Entries       map[string]map[string]string `json:"entries,omitempty"`
```
nit: a map to a map can be cumbersome, we can come back to this later with an actual type.

derekwaynecarr commented:
the breakdown per commit was really helpful, and the mechanical changes all made sense.

/approve
/lgtm

neolit123 commented:
@klueska i have a suspicion that this PR is making the 1.18 kubelet unstable, resulting in panics.

see this issue for more details:
https://github.com/kubernetes/kubernetes/issues/86676
